core/ledger/kvledger/txmgmt/validation/validator.go

```diff
@@ -78,38 +78,43 @@ func (v *validator) preLoadCommittedVersionOfRSet(blk *block) error {
 }
 
 // validateAndPrepareBatch performs validation and prepares the batch for final writes
-func (v *validator) validateAndPrepareBatch(blk *block, doMVCCValidation bool) (*publicAndHashUpdates, error) {
+func (v *validator) validateAndPrepareBatch(blk *block, doMVCCValidation bool) (*publicAndHashUpdates, []*AppInitiatedPurgeUpdate, error) {
 	// Check whether statedb implements BulkOptimizable interface. For now,
 	// only CouchDB implements BulkOptimizable to reduce the number of REST
 	// API calls from peer to CouchDB instance.
 	if v.db.IsBulkOptimizable() {
 		err := v.preLoadCommittedVersionOfRSet(blk)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 	}
 
 	updates := newPubAndHashUpdates()
+	purgeTracker := newPvtdataPurgeTracker()
+
 	for _, tx := range blk.txs {
 		var validationCode peer.TxValidationCode
 		var err error
 		if validationCode, err = v.validateEndorserTX(tx.rwset, doMVCCValidation, updates); err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 
 		tx.validationCode = validationCode
 		if validationCode == peer.TxValidationCode_VALID {
 			logger.Debugf("Block [%d] Transaction index [%d] TxId [%s] marked as valid by state validator. ContainsPostOrderWrites [%t]", blk.num, tx.indexInBlock, tx.id, tx.containsPostOrderWrites)
+
 			committingTxHeight := version.NewHeight(blk.num, uint64(tx.indexInBlock))
 			if err := updates.applyWriteSet(tx.rwset, committingTxHeight, v.db, tx.containsPostOrderWrites); err != nil {
-				return nil, err
+				return nil, nil, err
 			}
+
+			purgeTracker.update(tx.rwset, committingTxHeight)
```

Comment:
applyWriteSet() already iterates through all the write sets in the prior line.
purgeTracker.update() will iterate through all the write sets again.
It seems inefficient to iterate through write sets twice. But you probably did it this way to keep the code cleaner?

Reply:
Yes, that is correct. Also, we have seen that most of the transactions contain a handful of writes (typically one or two) so in practice, I will be surprised if we notice any measurable difference.